Bluetooth: Gatt settings fixes 2 electric boogaloo

jori-nordic opened the pull request:
See the commit messages. Mostly addresses #54770

Fixes #54770
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/54813

jori-nordic committed:
tests: Bluetooth: reproduce gatt settings bug

Reproduce #54470.

- client connects
- client writes to CF & CCC
- client bonds
- server reboots

After this, we would expect to have the CCC and CF values be correctly
restored from settings but they are not.

Additionally add the same test with BT_PRIVACY=y, which shows that the
feature is broken with that config (fixed in a subsequent commit).

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic committed:
Bluetooth: host: fix subtle bug in db_hash_process

Fixes a bug introduced in the previous refactoring: we would always mark
all peers as change-unaware. Now we only do so when the hash has been
recalculated.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic committed:
Bluetooth: host: save CF and CCC values written before bonding

On bond establishment: save the CF and CCC data that have been written
before the peer was bonded.

On identity resolved: update the CF data to use the peer's identity address
instead of its private address (same as is currently done for the CCC).

Fixes #54770.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

MarekPieta commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -878,6 +878,8 @@ static void db_hash_process(struct k_work *work)
 	if (!atomic_test_bit(gatt_sc.flags, DB_HASH_VALID)) {
 		db_hash_gen();
+	} else {
+		return;
```
If we generate the hash before settings are loaded (e.g. if a GATT service is registered before settings are loaded), we could accidentally skip setting `DB_HASH_LOAD_PROC` and performing the related operations needed after db_hash load from settings is completed.

MarekPieta replied:
btw. Shouldn't the condition below be `if (already_processed)` (without the negation) ?

jori-nordic replied:
ugh yeah. shouldn't have refactored this I guess

MarekPieta commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1496,14 +1510,24 @@ void bt_gatt_init(void)
+		.pairing_complete = bt_gatt_pairing_complete,
 	};
 
 	/* Register the gatt module for connection callbacks so it can be
```
`... for authentication info callbacks ...` ?

MarekPieta commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1119,6 +1119,20 @@ static void bt_gatt_identity_resolved(struct bt_conn *conn, const bt_addr_le_t *
+}
+
+static void bt_gatt_pairing_complete(struct bt_conn *conn, bool bonded)
+{
```
Shouldn't we skip storing the values if not bonded?

MarekPieta commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1119,6 +1119,20 @@ static void bt_gatt_identity_resolved(struct bt_conn *conn, const bt_addr_le_t *
 
 	bt_gatt_foreach_attr(0x0001, 0xffff, convert_to_id_on_match, &user_data);
 
+	/* Store the ccc */
```
We should make sure if the peer is bonded before storing data I think.

MarekPieta replied:
The cf_cfg in RAM should be updated anyway

MarekPieta commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1496,14 +1510,24 @@ void bt_gatt_init(void)
+	};
+
+	/* Also update the address of CCC or CF writes that happened before the
+	 * bond. Note that to increase security in the future, we might want to
```
`before the identity resolution`?

MarekPieta reviewed (approved):
One minor comment. Since it should not break anything, approved.

MarekPieta committed:
Bluetooth: host: Fix SC indication on encrypt change

The SC indication can be sent only if the GATT Service Changed is
enabled in the configuration. Change fixes build issues related to
implicit sc_indicate declaration when Service Changed is disabled.

Fixes: #54813

Signed-off-by: Marek Pieta <Marek.Pieta@nordicsemi.no>

MarekPieta commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1498,13 +1519,24 @@ void bt_gatt_init(void)
+	 * brand-new peer.
 	 */
 	bt_conn_cb_register(&gatt_conn_cb);
 #endif /* CONFIG_BT_GATT_CLIENT && CONFIG_BT_SETTINGS && CONFIG_BT_SMP */
```
We need to move the newly added functionality to use it also for cases when BT_GATT_CLIENT role is disabled. Otherwise it would have no effect at all for devices that do not support GATT Client role.

MarekPieta replied:
I missed this one initially, I noticed it making an additional test other Bluetooth Central.

MarekPieta replied:
Fixed here: https://github.com/jori-nordic/zephyr/pull/1

MarekPieta committed:
Bluetooth: host: Remove GATT Client dependency for storing CCC and CF

Change removes GATT Client dependency for storing CCC and CF on pairing
complete and identity resolved. These features are needed also if GATT
Client role is disabled.

Signed-off-by: Marek Pieta <Marek.Pieta@nordicsemi.no>

MarekPieta commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -876,7 +876,9 @@ static void sc_indicate(uint16_t start, uint16_t end);
 static void db_hash_process(struct k_work *work)
 {
-	if (!atomic_test_bit(gatt_sc.flags, DB_HASH_VALID)) {
+	bool new_hash = !atomic_test_bit(gatt_sc.flags, DB_HASH_VALID);
```
@jori-nordic, could you take a look at this one?

jori-nordic replied:
I think I've seen in my test that the run can be scheduled even if `DB_HASH_VALID` is set.

MarekPieta replied:
Discussed offline.
I think it was before hash was initially calculated after finished settings load. Still we can leave it as it is for now, because it should not break anything.

MarekPieta commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5748,7 +5748,7 @@ void bt_gatt_encrypt_change(struct bt_conn *conn)
 	bt_gatt_foreach_attr(0x0001, 0xffff, update_ccc, &data);
 
-#if defined(CONFIG_BT_SETTINGS)
+#if defined(CONFIG_BT_SETTINGS) && defined(CONFIG_BT_GATT_SERVICE_CHANGED)
```
Hi @jori-nordic
Seems that you (probably accidentally) cherry-picked the fix without comment alignment.

MarekPieta replied:
Discussed offline. Will be fixed separately

jori-nordic replied:
will do that in a subsequent PR to not block the releases.

jori-nordic commented:
@stephanosio can this be merged?

stephanosio commented:
@jori-nordic Is that last commit ("Bluetooth: host: Fix SC indication on encrypt change") necessary to fix #54770? If not, it should not be part of this series.

jori-nordic commented:
> @jori-nordic Is that last commit ("Bluetooth: host: Fix SC indication on encrypt change") necessary to fix #54770? If not, it should not be part of this series.

It can be, yes, because the whole series is useless if we don't select a kconfig (`BT_GATT_CLIENT`) that is not selected for most applications.

stephanosio reviewed (approved):
Merging since this series is fixing a release blocker issue.
